hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -694,6 +694,13 @@ class Emit[C](
             Class.forName("is.hail.expr.ir.package$"), "uuid4"))))
       case x@Literal(t, v) =>
         presentPC(mb.addLiteral(v, x.pType))
+      case x@EncodedLiteral(codec, value) =>
```

Comment:
clever, but we can't do this. This means that every usage of the literal will decode the encoded value -- think if about if we have:

```
StreamMap
  StreamRange idx 0 100 1
  ArrayRef 
    Ref idx
    EncodedLiteral(some array)
```

The literal will be decoded for every element.

I think the right solution is to extend the underlying infrastructure that encodeLiterals() calls to allow for an Array[Array[Byte]] that are decoded separately into separate fields on the class. This means that this line:

https://github.com/hail-is/hail/blob/f48f6fc56f8febcf6d1b87aa9ab89c43f67487d5/hail/src/main/scala/is/hail/expr/ir/EmitClassBuilder.scala#L318

Is going to have to decode an Array[Array[Byte]] into multiple fields, and so on.

Reply:
I don't quite understand yet. We are talking about decoding literals to separate fields, but then you talked about `encodeLiterals`. It seems like that method encodes all the literals currently stored in the `literalsMap`. I have an `[Array[Byte]]` that lives outside of that on the IR node. The current emit rule adds the `Array[Byte]` into that map, but encoding that would be weird, as it's already encoded.

Reply:
Oh! Is encoding a `PCanonicalBinary` a no-op, since it's already just bytes?

Reply:
I think I often don't communicate my design ideas effectively, and I'm trying to improve! Here's another shot:

encodeLiterals() takes the `literalsMap`, wraps it in a tuple, encodes it to bytes, broadcasts the bytes, and decodes the bytes on the worker.

The function we want to call to emit EncodedLiteral just should do the last two: broadcast the bytes, decode the bytes on the worker. This means I think we need to extend/abstract the machinery used here so that in addition to tracking a `literalsMap`, we also track encoded literals: `mutable.Map[mutable.Map[EncodedLiteral, PSettable]`. `encodeLiterals()` should return an `Array[Array[Byte]]` (say the 0th element is the current return value built literalsMap) with the 1-Nth elements being arrays taken directly from the EncodedLiterals. The code emitted into `mb2` in encodeLiterals() should also decode into the PSettables inside the encoded literals map.

Is this any clearer? Still not describing it super well.

Reply:
Totally clear, makes sense. Thanks!